### PR TITLE
Update RoslynAnalyzers to fix diff churn with PublicAPI.Unshipped.txt files

### PR DESCRIPTION
**Pull request by jcouv**
This PR does two things:
- Pulls in the latest RoslynAnalyzers package, so that we get the new behavior for PublicAPI fixer. The new behavior is that inserting a new API does not sort the whole "unshipped" file. The purpose is to minimize diff churn.
- Adds a batch script that sorts the "unshipped" files (may be useful to Tiger or for API review).

Fixes https://github.com/dotnet/roslyn/issues/26285

Tagging @jaredpar @dotnet/roslyn-infrastructure for review. Thanks

### CODE REVIEW

File: build/Rulesets/Roslyn_BuildRules.ruleset [hunk 1]
```diff
@@ -21,13 +21,16 @@
     <Rule Id="CA1055" Action="None" />
     <Rule Id="CA1056" Action="None" />
     <Rule Id="CA1061" Action="None" /> <!-- "do not hide base class methods": currently violations in the compiler -->
+    <Rule Id="CA1062" Action="None" />
```
Comment (jcouv) [inline build/Rulesets/Roslyn_BuildRules.ruleset]: @mavasani Let me know if I should open follow-up issues for the disabled rules I added. Or maybe some comment. Thanks

Comment (mavasani) [reply build/Rulesets/Roslyn_BuildRules.ruleset]: Yes, please do so and assign it to me. A single issue tracking enabling of all DFA based rules will suffice.

**Comment by jcouv** [comment]
@mavasani It looks like my local build didn't catch all. CI is showing RS1022 errors and a couple of RS0017 errors.
I assume RS1022 should be suppressed (Action=None) too. Is that right?
I'm investigating the RS0017 errors:

```
PublicAPI.Shipped.txt(4,1): error RS0017: Symbol 'Microsoft.CodeAnalysis.FileTextLoader.DefaultEncoding { get; } -> System.Text.Encoding (forwarded, contained in Microsoft.CodeAnalysis.Workspaces)' is part of the declared API, but is either not public or could not be found [C:\repos\roslyn\src\Workspaces\Core\Desktop\Workspaces.Desktop.csproj]
PublicAPI.Shipped.txt(7,1): error RS0017: Symbol 'Microsoft.CodeAnalysis.FileTextLoader.Path { get; } -> string (forwarded, contained in Microsoft.CodeAnalysis.Workspaces)' is part of the declared API, but is either not public or could not be found [C:\repos\roslyn\src\Workspaces\Core\Desktop\Workspaces.Desktop.csproj]
```

**Comment by jcouv** [comment]
@DustinCampbell Looking that the [recent history](https://github.com/dotnet/roslyn-analyzers/commits/master/src/Roslyn.Diagnostics.Analyzers/Core) for DeclaredAPI analyzer, it looks like you made a change related to forwarded properties.
I've removed them from shipped.txt file.
Can you review and confirm this is correct? Thanks

**Comment by mavasani** [comment]
> After further consideration, I believe that we should instead report a new warning if the file is not properly sorted.

@sharwell, that basically goes against the original intention of Julien's change - avoid requiring a fully sorted public API file to allow easy incremental reviews. A new diagnostic for unsorted file will break the build.

**Comment by sharwell** [comment]
@mavasani the problem with the previous approach was not the fact that the tool sorted the file. The problem was users who resolved merge conflicts in the file failed to preserve the relative sort order of additions, a situation which was only corrected the next time a change was made to the file. An analyzer during the build would ensure that merge conflicts were correctly resolved, thus addressing the incremental review situation equally well but leaving the file in a better overall state.

**Comment by jcouv** [comment]
I think that is workable. The analyzer can detect an unsorted file and tell you to run `script/build/sort-unshipped.cmd`.

**Comment by jcouv** [comment]
On the other hand, Aleksey made a good point: it is not common for people to look at this file aside from code reviews. Maintaining it sorted when sorting doesn't help in a common scenario is wasted effort.
@sharwell When do you think someone looks at this file? If more or less once per release, then I would suggest they run the script and check in the sorted files. There is no point in burdening Tiger with a frequent by low-value task.

**Comment by jcouv** [comment]
@mavasani Can you confirm that suppressing `RS1022` is also appropriate?
@DustinCampbell Can you take a look at the removed properties in "unshipped" file?

File: build/Rulesets/Roslyn_BuildRules.ruleset [hunk 2]
```diff
@@ -77,6 +82,7 @@
     <Rule Id="RS1012" Action="None" />
     <Rule Id="RS1013" Action="None" />
     <Rule Id="RS1014" Action="Warning" />
+    <Rule Id="RS1022" Action="None" /> <!-- https://github.com/dotnet/roslyn/issues/26400: Enable rules CA1062, CA1303, CA1508, CA2100 and CA2215 for Roslyn.sln -->
```
Comment (mavasani) [inline build/Rulesets/Roslyn_BuildRules.ruleset]: @jcouv `RS1022` is the new analyzer rule that flags analyzers that access any APIs from Workspace layer (DoNotUseTypesFromAssemblyRuleId). Given that lot of our IDE analyzers use backdoor access to Workspace options, we cannot turn it on until @sharwell is done moving all IDE analyzers to NuGet. Can you please file a separate issue to track enabling this rule and assign it to Sam?

**Comment by mavasani** [comment]
I like @sharwell's idea of keeping the file sorted, as long as there is no **manual** burden on anyone. @jcouv, if we can automate running your script as part of the tool that creates merge pull requests, then I am fine with enhancing the declare public API analyzer to report an additional file diagnostic for an unsorted file.

File: build/Rulesets/Roslyn_BuildRules.ruleset (hunk 2 above)
Comment (jcouv) [reply build/Rulesets/Roslyn_BuildRules.ruleset]: Filed https://github.com/dotnet/roslyn/issues/26420 and updated comment accordingly

**Comment by jcouv** [comment]
@jmarolf @jasonmalinowski I didn't hear back from you.
Would it be possible for tool that creates merges to run the `script/build/sort-unshipped.cmd` command as part of creating merges?

**Comment by KirillOsenkov** [comment]
Good times. For posterity I'd like to note that the sorting approach actually worked better if you sort both the source and the destination before the merge. If you require that the master is sorted and that a branch is also sorted before the merge, it does work as intended and simplifies things. If you run into conflicts because master sorted things, just sort your branch before attempting the merge and it'll auto-fix itself.

I do agree however that sorting of this type is just fighting the tools and this is always an uphill battle (swimming upstream?). The proper solution of course is to ensure that the tools preserve the ordering properly. Probably just as futile as trying to keep the line endings consistent and then using an editor that doesn't preserve line endings correctly.

I also note that the wildcards have made the whole thing a non-issue for SDK-style .csproj files, which is good.

**Comment by jcouv** [comment]
Thanks for the clarification.
Based on this, I don't think we have a reasonable solution to keep the file sorted during merges. So I'll go ahead with the PR as-is.
@sharwell I'll wait to hear from you before I merge. Thanks

**Comment by sharwell** [comment]
No objections to merge

**Pull request merged**